source/extensions/common/tap/tap_matcher.cc

```diff
@@ -121,6 +129,159 @@ void HttpHeaderMatcherBase::matchHeaders(const Http::HeaderMap& headers,
   statuses[my_index_].might_change_status_ = false;
 }
 
+// HttpGenericBodyMatcher
+// Scans the HTTP body and looks for patterns.
+// HTTP body may be passed to the matcher in chunks. The search logic buffers
+// only as many bytes as is the length of the longest pattern to be found.
+HttpGenericBodyMatcher::HttpGenericBodyMatcher(
```

Comment:
If you've got one long string and one small string, are we going to search the whole large buffer for the small string every time a byte comes in?

Reply:
At the moment logic is like this: if there are 2 patterns: 100 bytes long and 3 bytes long, the matcher buffers 99 bytes from the previous body when onBody is called. At the moment search for both patterns (100 bytes long one and 3 bytes long one) start from the beginning of the 99-bytes long buffer. I see two possible improvements here: 
- for 3 bytes long pattern the search may be started at offset 97
- once 100 bytes long pattern has been found, for subsequent searches only 2 bytes should be buffered.

Those are obvious low hanging fruits and I will implement those asap.

Reply:
Changed the logic to start search from an offset for patterns shorter than the longest one.